docs/developers/developer-reference/smart-contract-build-reference.md

```diff
@@ -0,0 +1,324 @@
+
+Not everything, though, can be performed here. Notably, macros cannot access data from other modules or crates, all processing is local to the current contract or module. Therefore, we need another mechanism for working with the complete contract data.
+
+### b. The (generated) ABI generator
+
+ABIs are a collection of metatada about the contract. To build an ABI, we also need the data from the modules. The module macros cannot be called from the contract macros (macros are run at compilation, we are not even sure that modules will need to be recompiled!). Modules, however can be called. That is why we are actually generating _ABI generator functions_ for each module, which can call one another to retrieve the composite picture.
+
+Note: The ABI generator comes as an implementation of trait [ContractAbiProvider](https://docs.rs/elrond-wasm/latest/elrond_wasm/contract_base/trait.ContractAbiProvider.html).
+
+### c. Meta crate: generating the ABI
+
+The next question is how will this function be called. Whenever we compile the WASM contracts, we also produce the ABIs in JSON format. Rust has something called build scripts, which get called _after_ compiling a project, but for reasons that will become apparent later, they are not powerful enough for our usecase.
+
+Therefore, we have decided to include an extra crate in each smart contract project. It is always found in the `meta` folder in a contract, and it handles the entire build. To minimize boilerplate, it always only contains one line, that simply defers execution to the framework:
+
+```rust
+fn main() {
+    elrond_wasm_debug::meta::perform::<my_contract_crate::AbiProvider>();
+}
+```
+
+To produce the ABI, in fact, it is enough to call:
+
+```sh
+cd meta
+cargo run
+```
+
+The meta crate has access to the ABI generator, because it always has a dependency to the contract crate. This is the `my_contract_crate::AbiProvider` in the example above.
+
+This is also the step where the meta crate parses and processes the `multicontract.toml` file. If there are multiple outputs, one ABI will be produced for each.
+
+
+### d. Meta crate: generating `wasm` crate code
+
+Each contract must contain at least one `wasm` crate. This is a separate crate from the contract crate, because it has a different purpose: it only needs to be the basis for compiling wasm. Take it as an intermediary step between the contract logic and the Rust to WASM compiler. This is also where the WASM compilation options are specified (e.g. the optimization level). These options can be seen in the the `Cargo.toml` file of the `wasm` crate.
+
+The separation is important, because it means that the smart contract crate can act as a pure rust crate, with no knowledge of WebAssembly. This makes it easy to test, coverage, and integrate in other unrelated technologies.
```

Comment:
```suggestion
The separation is necessary because, in this way, the smart contract crate can act as a pure rust crate with no knowledge of WebAssembly. This makes testing, coverage, and integration easy in other unrelated technologies.
```